webrtc/README.md

```diff
@@ -161,6 +165,50 @@ MUST pass an empty string. When receiving an `RTCDataChannel` via
 an empty string. This allows future versions of this specification to make use
 of the `RTCDataChannel` `label` property.
 
+## Closing an `RTCDataChannel`
+
+Some WebRTC implementations do not guarantee that any queued messages will be
+sent after a datachannel is closed.  Other implementatations maintain separate
+outgoing message and transport queues, the status of which may not be visible
+to the user. Consequently we must add an additional layer of signaling to ensure
+reliable data delivery.
+
+When a node wishes to close a stream for writing, it should send a message with
+the `FIN` flag set, then proceed to wait for a `FIN_ACK` message from the remote
+node.
+
+If a `FIN` flag is received and the node has finished writing data, it should
+respond with a `FIN_ACK` immediately.
+
+If a `FIN` flag is recieved but the node has not finished writing data, it should
+delay sending a `FIN_ACK` response until it has finished writing data, sent a
+`FIN` flag, and finally received a `FIN_ACK` in response.
+
+When a `FIN_ACK` has been sent and received, a node may close the datachannel.
+
+The node may close the datachannel without receiving a `FIN_ACK`, for example in
+the case of a timeout, but there will be no guarantee that all previously sent
+messages have been received.
+
+If a node has previously sent a `STOP_SENDING` flag to the remote node, it must
```

Comment:
```suggestion
If a node has previously sent a `STOP_SENDING` flag to the remote node before a FIN, it must
```

Just to make sure we are on the same page, one does not expect to receive a FIN_ACK after a STOP_SENDING, correct?

Reply:
The protobuf definition says about `STOP_SENDING`;

> The sender will no longer read messages on the stream. Incoming data is being discarded on receipt.

My reading of that was that the sender would not process any incoming messages at all after sending this flag but I was trying to call out that they should actually continue to process the flags from incoming messages (ignoring any stream data) in order to receive a `FIN_ACK` message.

Maybe it could be better worded?

Reply:
Also I think you could send `FIN` (e.g. close write) then `STOP_SENDING` (e.g. close read) before receiving a `FIN_ACK` - and adding the `before a FIN` in the suggestion above excludes this scenario?

Reply:
How about `before or after a FIN`? What I want to avoid is that a reader misunderstands that a `STOP_SENDING` is also acknowledged by a `FIN_ACK`.

Reply:
Maybe we could explicitly state that with something like:

> The FIN_ACK flag acknowledges the previous receipt of a message with the FIN flag set.
On receiving the FIN_ACK flag the recipient may now close the datachannel as they know the remote has received all data that was sent before the FIN.

It might be getting a bit wordy, what do you think to reducing it to just:

> Sending the FIN_ACK flag acknowledges the previous receipt of a message with the FIN flag set.

...and limiting any other exposition to the "Closing an `RTCDataChannel`" section?

Reply:
I've simplified the text here leaving the interpretation to the "Closing an RTCDataChannel" section.